### ISSUE TITLE
Module does not work with Ephemeral Agents/Runners

### ISSUE THREAD

**Issue by dansiviter** [terraform-google-modules/terraform-google-gcloud#94]
This module is not designed for ephemeral runners/agents such as Terraform Cloud Remote Operations that cannot have `gcloud` on the `PATH` and cannot have custom base images. If anything fails you have to unwind and re-apply all the changes again to force `gcloud` to be re-downloaded. 

Example:

```
module.asm_install.module.gcloud_kubectl.null_resource.additional_components[0] (local-exec): Executing; ["/bin/sh" "-c" ".terraform/modules/asm_install/scripts/check_components.sh .terraform/modules/asm_isntall/cache/xxxx/google-cloud-skd/bin/gcloud kubectl,kpt,beta.kustomize"]

Error: Error running command '.terraform/modules/asm_install/scripts/check_components.sh .terraform/modules/asm_isntall/cache/xxxx/google-cloud-skd/bin/gcloud kubectl,kpt,beta.kustomize': exit status 127. Output:
```

In the above example, due to a previously failed run the repeat run then fails to get passed this as it assumes `gcloud` exists in the cache. However, when it runs it returns a 127 command not found code.

**Comment by morgante** [comment]
Are you setting the `GCLOUD_TF_DOWNLOAD` variable? it is specifically to enable downloading in environments where `gcloud` is not present.

**Comment by dansiviter** [comment]
Apologies, I should have stated `skip_download: false` is set.

**Comment by morgante** [comment]
I see, I'm not actually sure how we can fix this though. Open to ideas.

**Comment by dansiviter** [comment]
I would argue we need to break the link between persisted TF state and local caching. That would mean dropping the following: 
* `random_id.cache`,
* `null_resource.prepare_cache`,
* `null_resource.download_gcloud`,
* `null_resource.download_jq`.

Then performing caching from within each script to check for `gcloud` (or `jq`) each time. This will maintain the exiting functionality running locally and caching the binaries. When this runs in an ephemeral agent/runner it will be slower as it'll be downloading each time but should at least it should complete (bar any download issues).

**Comment by morgante** [comment]
Moving the logic into each script (and, therefore, into bash) would probably work, though it does increase complexity.

However, I should be clear it's not necessarily a top priority for our team. In general we've found adding more functionality for downloading `gcloud` to add complexity (and a source of bugs) and are focusing more on having it available in all execution environments. So we're probably unlikely to work on this soon.

If this is a priority for you, I'd be happy to review a pull request though. I do think it's reasonable to move the check into each script.

**Comment by morgante** [comment]
We'd love to move ASM install away from a bash script, but it definitely won't be happening soon.

**Comment by AndreaGiardini** [comment]
I am hitting the same problem on terraform cloud. Have you found any workaround?

**Comment by onsails** [comment]
I've managed to work-around adding for_each to the module arguments:

```
module "gcloud" {
  # force new state on each apply
  for_each = {
    timestamp = "${timestamp()}"
  }

  ...
}
```

**Comment by trajan0x** [comment]
this is still breaking for me on ephemeral destroys

**Comment by Mithun2507** [comment]
in the recent upgrade of the terraform enterprise where we need to move to the new hashicorp/tfc-agent we are experiencing this issue. Do we have any fix on this from container image creation perspective? 
In the custom tfe agent image we are installing the gcloud SDK and adding the cloud SDK path to the global profile but still we get below error

Error: local-exec provisioner error
with module.gcloud-module-test.null_resource.run_destroy_command[0]
on .terraform/modules/gcloud-module-test/main.tf line 258, in resource "null_resource" "run_destroy_command":
  provisioner "local-exec" {
Error running command 'PATH=/root/.tfc-agent/component/terraform/runs/run-wRE74yf2dNFqyGmg/config/sandbox/.terraform/modules/gcloud-module-test/cache/380921a3/google-cloud-sdk/bin:$PATH
gcloud info
': exit status 127. Output: /bin/sh: 2: gcloud: not found

**Comment by bharathkkb** [comment]
@Mithun2507 in the custom TFE image is gcloud in path? Could you try locally pulling the image and checking if you can execute gcloud commands?

@peridotml Setting download to true should work but it would be more robust if you can create a custom image that has gcloud installed.

**Comment by Radiergummi** [comment]
I just hit this issue, with otherwise innocuous TF configuration. As this has been open for more than two years now, maybe it'd be time to either add a failure condition to the bash script if running in an ephemeral environment, or simply state in the documentation that the module isn't compatible with Terraform Cloud and alike? That would save users a bunch of time without lots of work....

**Comment by Mithun2507** [comment]
> @Mithun2507 in the custom TFE image is gcloud in path? Could you try locally pulling the image and checking if you can execute gcloud commands?

Yes, in the image we had the path specified. I tested the image locally and gcloud works fine locally. Hashi informed us they don't support gcloud in the new tfc-agent image

**Comment by anonimitoraf** [comment]
Just encountered this too. All this time I thought I was doing something wrong 😭

**Comment by imranzunzani** [comment]
This is still a problem. It will be helpful to skip referencing the cache by introducing a flag like `skip_cache=true` , which enforces `skip_download=false` every time.

**Comment by imranzunzani** [comment]
> I've managed to work-around adding for_each to the module arguments:
> 
> ```
> module "gcloud" {
>   # force new state on each apply
>   for_each = {
>     timestamp = "${timestamp()}"
>   }
> 
>   ...
> }
> ```
> 
> Of course, it works for idempotent create_cmd and special handling in delete_cmd

This doesn't work. Fails while destroying elements.
